apollo-router/src/plugins/external.rs

```diff
@@ -326,63 +475,271 @@ impl Plugin for ExternalPlugin {
+                                crate::graphql::Response::builder().errors(vec![Error {
+                                    message:format!("couldn't deserialize coprocessor output body: {error}"),
+                                    ..Default::default()
+                                }]).build()
+                            });
+
+                            subgraph::Response {
+                                response: http::Response::builder()
+                                    .status(code)
+                                    .body(graphql_response)?,
+                                context: request.context,
+                            }
+                        };
+                        return Ok(ControlFlow::Break(res));
+                    }
+
+                    // Finally, process our reply and act on the contents. Our processing logic is
+                    // that we replace "bits" of our incoming request with the updated bits if they
+                    // are present in our co_processor_output.
+
+                    let new_body: crate::graphql::Request = match co_processor_output.body {
+                        Some(value) => serde_json::from_value(value)?,
+                        None => body,
+                    };
+
+                    request.subgraph_request = http::Request::from_parts(parts, new_body);
+
+                    if let Some(context) = co_processor_output.context {
+                        request.context = context;
+                    }
+
+                    if let Some(headers) = co_processor_output.headers {
+                        *request.subgraph_request.headers_mut() = internalize_header_map(headers)?;
+                    }
+
+                    if let Some(uri) = co_processor_output.uri {
+                        *request.subgraph_request.uri_mut() = uri.parse()?;
+                    }
+
+                    Ok(ControlFlow::Continue(request))
+                }
+            })
+        });
+
+        let response_layer = (self.response != Default::default()).then_some({
+            let response_config = self.response.clone();
+
+            MapFutureLayer::new(move |fut| {
+                let http_client = http_client.clone();
+                let coprocessor_url = coprocessor_url.clone();
+                let response_config = response_config.clone();
+                let service_name = service_name.clone();
+
+                async move {
+                    let mut response: subgraph::Response = fut.await?;
+
+                    // Call into our out of process processor with a body of our body
+                    // First, extract the data we need from our response and prepare our
+                    // external call. Use our configuration to figure out which data to send.
+
+                    let (parts, body) = response.response.into_parts();
+                    let bytes = Bytes::from(serde_json::to_vec(&body)?);
```

Comment:
Do you think it'd be worth having a dedicated error type for this plugin. The risk is that say the co-processor returns garbage. The errors in the logs will be won't give an indication of why things went wrong.
Something along the lines of `could not read response from co-processor` and then later on when serde happens `co-processor returned malformed response`

Reply:
So I had a quick go, and the error message in the logs for if the co-processor can't be contacted is:
```2023-02-21T11:27:09.648580Z  ERROR [trace_id=89c32d581dd59ed09bddc09d4f043454] response failed classification=Status code: 500 Internal Server Error latency=1 ms```
This should be more specific about what is wrong.

Reply:
Right, let's map_err and wrap it with an actionable message.